dix/req.c

```diff
@@ -1625,24 +1629,28 @@ static void dix_rop_completed(struct m0_sm_group *grp, struct m0_sm_ast *ast)
 	struct m0_dix_rop_ctx *rop_del_phase2 = NULL;
 	bool                   del_phase2 = false;
 	struct m0_dix_cas_rop *cas_rop;
+	int                    successful_ops = 0;
 
 	(void)grp;
 	if (req->dr_type == DIX_NEXT)
 		m0_dix_next_result_prepare(req);
 	else {
+		successful_ops = m0_tl_reduce(cas_rop, scan, &rop->dg_cas_reqs, 0,
+				  + !!(scan->crp_creq.ccr_sm.sm_rc == 0));
+
 		/*
-		 * Consider DIX request to be successful if there is at least
-		 * one successful CAS request.
+		 * If enough operations succeeded to satisfy min_replicas,
+		 * ignore any operations that failed. If we don't have enough,
+		 * count all operations when deciding overall success (ensures the
+		 * overall operation is considered a failure).
 		 */
-		if (m0_tl_forall(cas_rop, cas_rop,
-				 &rop->dg_cas_reqs,
-				 cas_rop->crp_creq.ccr_sm.sm_rc != 0))
-			    dix_cas_rop_rc_update(cas_rop_tlist_tail(
-						  &rop->dg_cas_reqs), 0);
 
+		M0_ASSERT(req->dr_min_replicas > 0);
 		m0_tl_for (cas_rop, &rop->dg_cas_reqs, cas_rop) {
-			if (cas_rop->crp_creq.ccr_sm.sm_rc == 0)
+			if (successful_ops < req->dr_min_replicas ||
```

Comment:
At least two cases are not handled properly:
* In case of all failed CAS items the rc of base DIX item is not updated and remain is 0
* In case of several successful CAS items for DIX GET seems like a memory leak occurs... inside of _dix_cas_rop_rc_update()_ in case of rc == 0 the function _dix_item_rc_update()_ is called where the pointer to the result is set and corresponding AT buffer is zeroed (to avoid freeing during finalization of AT buffer), during processing of the next CAS GET reply the pointer to the result is overwritten and previous pointer is lost. In fact the only last CAS GET result is used and all previous results are leaked.

```
		switch (rtype) {
		case DIX_GET:
			m0_cas_get_rep(creq, key_idx, &get_rep);
			rc = get_rep.cge_rc;
			if (rc == 0) {
				ditem->dxi_val = get_rep.cge_val;
				/* Value will be freed at m0_dix_req_fini(). */
				m0_cas_rep_mlock(creq, key_idx);
			}
			break;
```

Reply:
To solve the first issue a piece of removed code needs to be brought back
```
		if (m0_tl_forall(cas_rop, cas_rop,
				 &rop->dg_cas_reqs,
				 cas_rop->crp_creq.ccr_sm.sm_rc != 0))
			    dix_cas_rop_rc_update(cas_rop_tlist_tail(
						  &rop->dg_cas_reqs), 0);
```

Reply:
Proposal for solving of the second issue:
we need a logical change in _dix_item_rc_update()_. In DIX_GET case we need to check whether _rc == 0_ and the DIX item value buffer has already been set (_ditem->dxi_val_), if so then we need to compare versions (can be mocked as _if (1 /* versions: latest result wins*/)_), if the version of next result wins then free current DIX value buffer and set a new one.

Reply:
Yes, you're right about the first point, we'll get _(1 || 0)_ for all replies and update will be executed, missed that condition.

Reply:
I added a placeholder function to compare versions, along with a comment on what's going on. I think the leak in `dix_item_rc_update()` should be fixed now.